proposals/record-structs.md

```diff
@@ -0,0 +1,303 @@
+(accessibility on type, modifiers on members, parameterless instance constructors,
+`base(...)` instance constructor initializers, definite assignment for `this` in constructor, destructors, ...).
+
+See https://github.com/dotnet/csharplang/blob/master/spec/structs.md
+
+But instance field declarations for a record struct are permitted to include variable initializers when there is a primary constructor.
+(TODO: need to confirm, since default values still possible)
+
+At most one partial type declaration of a partial record struct may provide a `parameter_list`.
+The `parameter_list` may not be empty.
+
+Record struct parameters cannot use `ref`, `out` or `this` modifiers (but `in` and `params` are allowed).
+
+## Members of a record struct
+
+In addition to the members declared in the record struct body, a record struct type has additional synthesized members.
+Members are synthesized unless a member with a "matching" signature is declared in the record struct body or
+an accessible concrete non-virtual member with a "matching" signature is inherited.
+Two members are considered matching if they have the same
+signature or would be considered "hiding" in an inheritance scenario.
+
+It is an error for a member of a record struct to be named "Clone".
+
+It is an error for an instance field of a record struct to have an unsafe type.
+
+A record struct is not permitted to declare a destructor.
+
+The synthesized members are as follows:
+
+### Equality members
+
+The synthesized equality members are similar as in a record class (`Equals` for this type, `Equals` for `object` type, `==` and `!=` operators for this type),\
+except for the lack of `EqualityContract`, null checks or inheritance.
+
+The record struct implements `System.IEquatable<R>` and includes a synthesized strongly-typed overload of `Equals(R other)` where `R` is the record struct.
+The method is `public`.
+The method can be declared explicitly. It is an error if the explicit declaration does not match the expected signature or accessibility.
+
+If `Equals(R other)` is user-defined (not synthesized) but `GetHashCode` is not, a warning is produced.
+
+```C#
+public bool Equals(R other);
+```
+
+The synthesized `Equals(R)` returns `true` if and only if for each instance field `fieldN` in the record struct
+the value of `System.Collections.Generic.EqualityComparer<TN>.Default.Equals(fieldN, other.fieldN)` where `TN` is the field type is `true`.
+
+The record struct includes synthesized `==` and `!=` operators equivalent to operators declared as follows:
+```C#
+public static bool operator==(R r1, R r2)
+    => r1.Equals(r2);
+public static bool operator!=(R r1, R r2)
+    => !(r1 == r2);
+```
+The `Equals` method called by the `==` operator is the `Equals(R other)` method specified above. The `!=` operator delegates to the `==` operator. It is an error if the operators are declared explicitly.
```

Comment:
out of curiosity, why?  seems to go against our view that we'll defer to the user if they provide it.

Reply:
If you mean why it is an error for declaring those operators explicitly, it's for consistency with record class types.

---
In reply to: [553562440](https://github.com/dotnet/csharplang/pull/4307#discussion_r553562440) [](ancestors = 553562440)